lib/net.c

```diff
@@ -1789,20 +1972,28 @@ drop_privileges(void)
 		return -1;
 	}
 
-	if (initgroups(pw->pw_name, config.pw_gid) != 0) {
-		G_LOG(ERR, "%s: failed to call initgrous(%s, %u) - %s\n",
-			__func__, pw->pw_name, config.pw_gid, strerror(errno));
+	G_LOG(DEBUG, "Ambient capabilities supported: %s\n",
+		CAP_AMBIENT_SUPPORTED() ? "yes" : "no");
+
+	log_proc_caps(NULL, "Capabilities before changing privileges");
+
+	ret = set_groups(pw->pw_name, config.pw_gid);
+	if (ret < 0) {
+		G_LOG(ERR, "%s: failed to set groups for user %s (gid %d)\n",
+			__func__, pw->pw_name, config.pw_gid);
 		return -1;
 	}
 
-	/* Drop privileges. */
-	if (setgid(config.pw_gid) != 0 || setuid(config.pw_uid) != 0) {
-		G_LOG(ERR, "%s: failed to drop privileges for user with user id %u and group id %u: %s\n",
-			__func__, config.pw_uid,
-			config.pw_gid, strerror(errno));
-		return -1;
+	log_proc_caps(NULL, "Capabilities after changing group(s)");
+
+	ret = cap_setuid(config.pw_uid);
+	if (ret != 0) {
+		G_LOG(ERR, "%s: failed to set UID for user %s (uid %d): %s\n",
+			__func__, pw->pw_name, config.pw_uid, strerror(errno));
 	}
 
+	log_proc_caps(NULL, "Capabilities after changing user");
+
 	/* Sanity check to ensure process can't effectively be root. */
 	RTE_VERIFY(seteuid(0) == -1 && setegid(0) == -1);
```

Comment:
We must break this test into two tests. Each test should log and return an error when they fail. The reason we cannot use `RTE_VERIFY()` is that, consider the case in which the Gatekeeper user is in a group that allows a user to turn root (or another pathway to become root), we need to report the security flaw instead of failing. To test this issue, one can set `config.pw_gid` to zero. Thus, our sanity check is actually verifying the setup on the host, not only the correctness of our code.